src/Akka.Streams.Kafka/Stages/Consumers/Abstract/BaseSingleSourceLogic.cs

```diff
@@ -6,45 +6,56 @@
 using System.Threading.Tasks;
 using Akka.Actor;
 using Akka.Dispatch;
+using Akka.Streams.Kafka.Helpers;
 using Akka.Streams.Kafka.Settings;
 using Akka.Streams.Kafka.Stages.Consumers.Actors;
 using Akka.Streams.Kafka.Stages.Consumers.Exceptions;
 using Akka.Streams.Stage;
 using Akka.Streams.Supervision;
+using Akka.Streams.Util;
 using Confluent.Kafka;
 using Decider = Akka.Streams.Supervision.Decider;
 using Directive = Akka.Streams.Supervision.Directive;
 
 namespace Akka.Streams.Kafka.Stages.Consumers.Abstract
 {
     /// <summary>
-    /// Shared GraphStageLogic for <see cref="SingleSourceStageLogic{K,V,TMessage}"/> and <see cref="ExternalSingleSourceLogic"/>
+    /// Shared GraphStageLogic for <see cref="SingleSourceStageLogic{K,V,TMessage}"/> and <see cref="ExternalSingleSourceLogic{K,V,TMessage}"/>
     /// </summary>
     /// <typeparam name="K">Key type</typeparam>
     /// <typeparam name="V">Value type</typeparam>
     /// <typeparam name="TMessage">Message type</typeparam>
     internal abstract class BaseSingleSourceLogic<K, V, TMessage> : GraphStageLogic
     {
         private readonly SourceShape<TMessage> _shape;
-        private readonly TaskCompletionSource<NotUsed> _completion;
         private readonly IMessageBuilder<K, V, TMessage> _messageBuilder;
         private int _requestId = 0;
         private bool _requested = false;
         private readonly Decider _decider;
 
         private readonly ConcurrentQueue<ConsumeResult<K, V>> _buffer = new ConcurrentQueue<ConsumeResult<K, V>>();
         protected IImmutableSet<TopicPartition> TopicPartitions { get; set; } = ImmutableHashSet.Create<TopicPartition>();
+
+        /// <summary>
+        /// Implements <see cref="IControl"/> for logic control
+        /// </summary>
+        protected readonly PromiseControl<TMessage> InternalControl;
         
         protected StageActor SourceActor { get; private set; }
         internal IActorRef ConsumerActor { get; private set; }
+
+        /// <summary>
+        /// Implements <see cref="IControl"/> to provide control over executed source
+        /// </summary>
+        public virtual IControl Control => InternalControl;
```

Comment:
Why have this be `virtual` but `InternalControl` be `readonly`? Is the idea that you can compose different `IControl` instances on top of each other?

Reply:
In scala, every stage can override `PromiseControl`'s methods - so every stage can have it own implementation of `IControl` (usually overriding `PromiseControl`).
So `Control` is virtual, because we need to be able to set another `PromiseControl` inherited implementation as a control materialized value. And `InternalControl` is used internally because sometimes we need to call some `PromiseControl` methods, but stage exposes only `IControl`.

But you are right, actually - in scala stage uses mixins, but in general it exposes `PromiseControl` instance, not generic `IControl`... This will simplify things. I will update PR.